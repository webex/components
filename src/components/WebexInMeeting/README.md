# Webex In-Meeting Component

Webex In-Meeting component displays the [Webex Local Media](../WebexLocalMedia) component on top of the [Webex Remote Media](../WebexRemoteMedia) component for the in meeting media experience.

<p align="center">
  <img src="./WebexInMeeting.png" alt="Default Webex InMeeting" />
</p>

## Preview

To see all the different possible states of the Webex Interstitial Meeting component,
you can view our [storybook](https://webex.github.io/components/storybook/?path=/story/webex-inmeeting--remote-and-local-media-enabled)
or run the following **NPM** command:

```shell
  npm start
```

## Embed

1.  Create a component adapter from which the data will be retrieved (See [adapters](../../adapters)). For instance:

    ```js
    const jsonAdapter = new WebexJSONAdapter(jsonData);
    ```

2.  Create a component instance by passing the meeting ID as a string and
    enclose it within [a data provider](../WebexDataProvider/WebexDataProvider.js)
    that takes the [component data adapter](../../adapters/WebexJSONAdapter.js) that we created previously

    ```js
    <WebexDataProvider adapter={jsonAdapter}>
      <WebexInMeeting meetingID="meetingID" />
    </WebexDataProvider>
    ```

The component knows how to manage its data. If anything changes in the data source that the adapter manages, the component will also update on its own.
