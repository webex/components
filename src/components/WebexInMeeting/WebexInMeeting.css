.in-meeting {
  position: relative;
  width: 100%;
  height: 100%;
}

.in-meeting .remote-media {
  position: relative;
  width: 100%;
  height: 100%;
}

.in-meeting .local-media {
  position: absolute;
  bottom: 25%;
  right: 3%;
  width: 20%;
  height: 20%;
  z-index: 1;
}
