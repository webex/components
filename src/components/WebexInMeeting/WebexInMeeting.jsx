import React, {useContext, useEffect, useState} from 'react';
import PropTypes from 'prop-types';

import {MeetingState} from '@webex/component-adapter-interfaces';
import Banner from '../generic/Banner/Banner';
import WebexLocalMedia from '../WebexLocalMedia/WebexLocalMedia';
import WebexRemoteMedia from '../WebexRemoteMedia/WebexRemoteMedia';
import webexComponentClasses from '../helpers';
import {
  AdapterContext,
  useElementDimensions,
  useMeeting,
  useRef,
} from '../hooks';
import {TABLET, DESKTOP, DESKTOP_LARGE} from '../breakpoints';
import {Modal} from '../generic';
import WebexMeetingGuestAuthentication from '../WebexMeetingGuestAuthentication/WebexMeetingGuestAuthentication';
import WebexMeetingHostAuthentication from '../WebexMeetingHostAuthentication/WebexMeetingHostAuthentication';
import WebexMeetingInfo from '../WebexMeetingInfo/WebexMeetingInfo';
/**
 * Webex In-Meeting component displays the remote stream plus
 * the local stream at the bottom right corner.
 *
 * @param {object} props  Data passed to the component
 * @param {string} props.className  Custom CSS class to apply
 * @param {string} props.layout  Layout to apply on remote video
 * @param {string} props.meetingID  ID of the meeting for which to show media
 * @param {object} props.style  Custom style to apply
 * @param {object} props.setEscapedAuthentication Set true if user escapes authentication
 * @returns {object} JSX of the component
 */
export default function WebexInMeeting({
  className, layout, meetingID, style, setEscapedAuthentication,
}) {
  const adapter = useContext(AdapterContext);
  const {
    remoteShare,
    localShare,
    passwordRequired,
    state,
  } = useMeeting(meetingID);
  const meetingRef = useRef();
  const {width, height} = useElementDimensions(meetingRef);
  const [maxWidth, setMaxWidth] = useState('none');
  const [authModal, setAuthModal] = useState('guest');
  const localMediaType = localShare?.stream ? 'screen' : 'video';
  const [cssClasses, sc] = webexComponentClasses('in-meeting', className, {
    tablet: width >= TABLET && width < DESKTOP,
    desktop: width >= DESKTOP && width < DESKTOP_LARGE,
    'desktop-xl': width >= DESKTOP_LARGE,
    'remote-sharing': remoteShare !== null,
  });
  const {NOT_JOINED} = MeetingState;

  useEffect(() => {
    setMaxWidth(height ? (height * 16) / 9 : 'none');
  }, [height]);

  const onCloseHandler = () => {
    adapter.meetingsAdapter.clearPasswordRequiredFlag(meetingID);
    setEscapedAuthentication(true);
  };

  return (
    <div ref={meetingRef} className={cssClasses} style={style}>
      {passwordRequired && state === NOT_JOINED && (
      <Modal
        onClose={onCloseHandler}
        otherClassName={[sc('authentication')]}
        onBack={authModal === 'host' && (() => setAuthModal('guest'))}
        ariaLabel={authModal === 'guest' ? 'Meeting guest authentication' : 'Meeting host authentication'}
      >
        {
          authModal === 'guest'
            ? (
              <WebexMeetingGuestAuthentication
                meetingID={meetingID}
                className={sc('authentication-guest')}
                switchToHostModal={() => setAuthModal('host')}
                style={style}
              />
            )
            : <WebexMeetingHostAuthentication meetingID={meetingID} className={sc('authentication-host')} />
        }
      </Modal>
      )}
      <div style={{maxWidth}} className={sc('media-container')}>
        <WebexMeetingInfo className={sc('info')} meetingID={meetingID} />
        <WebexRemoteMedia className={sc('remote-media-in-meeting')} layout={layout} meetingID={meetingID} />
        <WebexLocalMedia className={sc('local-media-in-meeting')} meetingID={meetingID} mediaType={localMediaType} />
        {localShare?.stream && <Banner className={sc('share-banner')}>You&apos;re sharing your screen</Banner>}
      </div>
    </div>
  );
}

WebexInMeeting.propTypes = {
  className: PropTypes.string,
  layout: PropTypes.string,
  meetingID: PropTypes.string.isRequired,
  style: PropTypes.shape(),
  setEscapedAuthentication: PropTypes.func,
};

WebexInMeeting.defaultProps = {
  className: '',
  layout: undefined,
  style: undefined,
  setEscapedAuthentication: undefined,
};
