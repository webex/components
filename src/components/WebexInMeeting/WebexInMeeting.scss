$shadow-offset: 0.25rem;
$shadow-blur: 1rem;
$shadow-spread: -0.125rem;
$shadow-color: rgba(0, 0, 0, 0.2);

.#{$WEBEX_COMPONENTS_CLASS_PREFIX}-in-meeting {
  position: relative;
  height: 100%;
  width: 100%;

  background: black;

  &.remote-sharing {
    .local-media-in-meeting {
      display: none;
    }
  }

  .local-media-in-meeting {
    @include media-preview-mobile();

    min-width: auto;
    min-height: auto;

    .md-avatar {
      transform: none; // Avatar is 40x40px, resets any transforms
    }
  }

  &-tablet {
    .local-media-in-meeting {
      @include media-preview-tablet();

      .md-avatar {
        transform: scale(2); // Avatar is 40x40px, this makes it 80x80px
      }
    }
  }

    &-desktop {
    .local-media-in-meeting {
      @include media-preview-tablet();

      .md-avatar {
        transform: scale(2); // Avatar is 40x40px, this makes it 120x120px
      }
    }
  }

  &-desktop-xl {
    .local-media-in-meeting {
      @include media-preview-desktop-xl();

      .md-avatar {
        transform: scale(3); // Avatar is 40x40px, this makes it 120x120px
      }
    }
  }
}
