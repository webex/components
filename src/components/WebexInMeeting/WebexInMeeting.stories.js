import React from 'react';
import {storiesOf} from '@storybook/react';

import jsonData from '../../data';
import {WebexJSONAdapter} from '../../adapters';

import {WebexInMeeting, WebexDataProvider} from '..';

const stories = storiesOf('Webex InMeeting', module);
const webexAdapter = new WebexJSONAdapter(jsonData);

stories.add('remote and local media enabled', () => (
  <WebexDataProvider adapter={webexAdapter}>
    <WebexInMeeting meetingID="remote&localMedia" />
  </WebexDataProvider>
));

stories.add('only remote media enabled', () => (
  <WebexDataProvider adapter={webexAdapter}>
    <WebexInMeeting meetingID="remoteMedia" />
  </WebexDataProvider>
));
