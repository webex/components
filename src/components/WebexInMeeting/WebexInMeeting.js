import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import {WEBEX_COMPONENTS_CLASS_PREFIX} from '../../constants';
import {WebexLocalMedia, WebexRemoteMedia} from '../';
import {useElementDimensions} from '../hooks';
import {TABLET, DESKTOP} from '../breakpoints';

import './WebexInMeeting.scss';

/**
 * Webex In-Meeting component displays the remote stream plus
 * the local stream at the bottom right corner.
 *
 * @param {string} props.meetingID  ID of the meeting for which to show media
 * @returns {Object} JSX of the component
 */
export default function WebexInMeeting({meetingID}) {
  const [meetingRef, {width}] = useElementDimensions();
  const classBaseName = 'in-meeting';
  const classObjects = {};

  classObjects[`${WEBEX_COMPONENTS_CLASS_PREFIX}-${classBaseName}`] = true;
  classObjects[`${WEBEX_COMPONENTS_CLASS_PREFIX}-${classBaseName}-tablet`] = width >= TABLET && width < DESKTOP;
  classObjects[`${WEBEX_COMPONENTS_CLASS_PREFIX}-${classBaseName}-desktop`] = width >= DESKTOP;

  const cssClasses = classNames(classObjects);

  return (
    <div ref={meetingRef} className={cssClasses}>
      <WebexRemoteMedia meetingID={meetingID} />
      <WebexLocalMedia meetingID={meetingID} />
    </div>
  );
}

WebexInMeeting.propTypes = {
  meetingID: PropTypes.string.isRequired,
};
