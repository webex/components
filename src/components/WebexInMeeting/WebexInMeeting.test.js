import React from 'react';

import WebexInMeeting from '../WebexInMeeting/WebexInMeeting';

jest.mock('../hooks/useMeeting');

describe('Webex InMeeting component', () => {
  describe('snapshot', () => {
    test('matches snapshot of meeting', () => {
      expect(shallow(<WebexInMeeting meetingID="remoteMedia" />)).toMatchSnapshot();
    });
  });
});
