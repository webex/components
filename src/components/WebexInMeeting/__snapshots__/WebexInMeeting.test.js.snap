// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Webex InMeeting component snapshot matches snapshot of meeting 1`] = `
<div
  className="wxc-in-meeting"
>
  <WebexRemoteMedia
    className="remote-media-in-meeting"
    meetingID="remoteMedia"
  />
  <WebexLocalMedia
    className="local-media-in-meeting"
    meetingID="remoteMedia"
  />
</div>
`;
